components/brave_ads/core/internal/common/url/url_util_internal.cc

```diff
@@ -0,0 +1,210 @@
+    // Do not support hosts other than chrome://settings.
+    return false;
+  }
+
+  if (modified_url.path() == kSearchEnginesPath ||
+      modified_url.path() == kSearchPath) {
+    if (!modified_url.has_query()) {
+      // Support chrome://settings/searchEngines and chrome://settings/search
+      // paths without a query.
+      return true;
+    }
+
+    return HasSearchQuery(modified_url);
+  }
+
+  // We can reject everything else!
+  return false;
+}
+
+std::optional<size_t> GetRegistryLength(const GURL& url) {
+  // `EXCLUDE_PRIVATE_REGISTRIES` is intentional because we want to return
+  // only the registry, not the private registry.
+  const size_t registry_length =
+      net::registry_controlled_domains::GetRegistryLength(
+          url, net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES,
+          net::registry_controlled_domains::EXCLUDE_PRIVATE_REGISTRIES);
+  if (registry_length == std::string::npos || registry_length == 0) {
+    return std::nullopt;  // No registry.
+  }
+
+  return registry_length;
+}
+
+std::optional<std::string> GetDomain(const GURL& url) {
+  if (!url.is_valid() || !url.has_host() || url.HostIsIPAddress()) {
+    return std::nullopt;  // No registry.
+  }
+
+  const std::optional<size_t> registry_length = GetRegistryLength(url);
+  if (!registry_length) {
+    return std::nullopt;  // No registry.
+  }
+
+  const std::string host = url.host();
+
+  // Remove the registry from the host.
+  const size_t domain_length =
+      host.length() - (*registry_length + /*'.' before registry*/ 1);
+  CHECK_GE(host.length(), domain_length);
+  return host.substr(0, domain_length);
+}
+
+std::optional<std::string> GetRegistryControlledDomain(const GURL& url) {
+  if (!url.is_valid() || !url.has_host() || url.HostIsIPAddress()) {
+    return std::nullopt;  // No registry.
+  }
+
+  const std::string_view host_piece = url.host_piece();
+
+  if (!net::registry_controlled_domains::HostHasRegistryControlledDomain(
+          host_piece,
+          net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES,
+          net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES)) {
+    return std::nullopt;
+  }
+
+  const std::string domain_and_registry =
+      net::registry_controlled_domains::GetDomainAndRegistry(
+          host_piece,
+          net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
+  if (domain_and_registry.empty()) {
+    return std::nullopt;  // No domain.
+  }
+
+  const std::optional<size_t> registry_length = GetRegistryLength(url);
+  if (!registry_length) {
+    return std::nullopt;  // No registry.
+  }
+
+  // Remove the registry from the host.
+  const size_t domain_length = domain_and_registry.length() -
+                               (*registry_length + /*'.' before registry*/ 1);
+  CHECK_GE(domain_and_registry.length(), *registry_length);
+  return domain_and_registry.substr(0, domain_length);
+}
+
+bool IsHostOnPublicSuffixList(const GURL& url) {
```

Comment:
can we just call https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.h;l=182?q=GetDomainAndRegistry&ss=chromium and make sure it's nonempty here? 

IIUC what we are trying to check is that the domain is a valid host and is NOT itself an eTLD (like https://co.uk), and i think GetDomainAndRegistry does that.

it would return nonempty for a URL like https://foo.bar where the TLD is not on the PSL but i think that's ok.

Reply:
if we are trying to exclude non-PSL domains we could just check [HostHasRegistryControlledDomain](https://source.chromium.org/chromium/chromium/src/+/main:net/base/registry_controlled_domains/registry_controlled_domain.cc;drc=90cac1911508d3d682a67c97aa62483eb712f69a;bpv=1;bpt=1;l=429?gsn=HostHasRegistryControlledDomain&gs=KYTHE%3A%2F%2FKoYDCsABa3l0aGU6Ly9jaHJvbWl1bS5nb29nbGVzb3VyY2UuY29tL2NvZGVzZWFyY2gvY2hyb21pdW0vc3JjLy9tYWluP2xhbmc9YyUyQiUyQj9wYXRoPW5ldC9iYXNlL3JlZ2lzdHJ5X2NvbnRyb2xsZWRfZG9tYWlucy9yZWdpc3RyeV9jb250cm9sbGVkX2RvbWFpbi5jYyNGVWZuR0NDbDQ1ZGVCMXJiRHcyQ0Y2VzFHbGRZaTVOMHFJbjVKU3Z6dkFJCsABa3l0aGU6Ly9jaHJvbWl1bS5nb29nbGVzb3VyY2UuY29tL2NvZGVzZWFyY2gvY2hyb21pdW0vc3JjLy9tYWluP2xhbmc9YyUyQiUyQj9wYXRoPW5ldC9iYXNlL3JlZ2lzdHJ5X2NvbnRyb2xsZWRfZG9tYWlucy9yZWdpc3RyeV9jb250cm9sbGVkX2RvbWFpbi5jYyNxSk5uTWV6SXU2b2VDaENCSzVHTGw0cl9yellORTlHVHhKZTlsaGVhdWEw) i think?

Reply:
```
net::registry_controlled_domains::HostHasRegistryControlledDomain("example.mysite.appspot.com"), net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `true` because `mysite.appspot.com` is a partial match.
```
net::registry_controlled_domains::GetDomainAndRegistry("example.mysite.appspot.com"),
net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `mysite.appspot.com`.

But we do not want to support `example.mysite.appspot.com`.

Reply:
Another example with www.google.com

```
net::registry_controlled_domains::HostHasRegistryControlledDomain("google.com"), net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `true` because `google.com` is a match.
```
net::registry_controlled_domains::GetDomainAndRegistry("google.com"),
net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `google.com`.

Reply:
So, @fmarier @diracdeltas to not support `example.mysite.appspot.com` I needed to add additional logic.

Reply:
Another example:

```
net::registry_controlled_domains::HostHasRegistryControlledDomain("foo.bar.fred.waldo.com"), net::registry_controlled_domains::EXCLUDE_UNKNOWN_REGISTRIES, net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `true` because `waldo.com` is a match.
```
net::registry_controlled_domains::GetDomainAndRegistry("foo.bar.fred.waldo.com"),
net::registry_controlled_domains::INCLUDE_PRIVATE_REGISTRIES);
```
returns `waldo.com`.

Reply:
Sorry, I misread the intention of that function.
We do want to support URLs that are ETLD+2 like `foo.bar.appspot.com` (or `www.google.com`, `www.google.co.uk`).

Reply:
Thanks. So ETLD+1 or more.

Reply:
This code has now been removed :-)